Rubberduck.Parsing/UIContext/UiSynchronizer.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Rubberduck.Parsing.UIContext
+{
+    public static class UiSynchronizer
+    {
+        private enum DllVersion
+        {
+            Unknown,
+            Vbe6,
+            Vbe7
+        }
+
+        private static DllVersion _version;
+
+        private static SynchronizationContext UiContext { get; set; }
+
+        static UiSynchronizer()
+        {
+            _version = DllVersion.Unknown;
+        }
+
+        public static void Initialize()
+        {
+            if (UiContext == null)
+            {
+                UiContext = SynchronizationContext.Current;
+            }
+        }
+
+        /// <summary>
+        /// Used to pump any pending COM messages. This should be used only as a part of
+        /// synchronizing or to effect a block until all other threads has finished with 
+        /// their pending COM calls. This should be used by the UI thread **ONLY**; 
+        /// otherwise execptions will be thrown.
+        /// </summary>
+        /// <remarks>
+        /// Typical use would be within a event handler for an event belonging to a COM 
+        /// object which require some synchronization with COM accesses from other threads. 
+        /// Events raised by COM are on UI thread by definition so the call stack originating
+        /// from COM objects' events can use this method.
+        /// </remarks>
+        /// <returns>Count of open forms which is always zero for VBA hosts but may be nonzero for VB6 projects.</returns>
+        public static int DoEvents()
+        {
+            CheckContext();
+            
+            return ExecuteDoEvents();
+        }
+
+        private static int ExecuteDoEvents()
+        {
+            switch (_version)
+            {
+                case DllVersion.Vbe7:
+                    return rtcDoEvents7();
+                case DllVersion.Vbe6:
+                    return rtcDoEvents6();
+                default:
+                    return DetermineVersionAndExecute();
+            }
+        }
+
+        private static int DetermineVersionAndExecute()
```

Comment:
My concern is that because it's a `static` class with a `static` ctor, it's hard to control when it gets initialized and exceptions from a static ctor are particularly troublesome to debug. It is also a micro-optimization because the `switch` will only execute the function after it has considered all other cases, which is better for more typical case where function is re-called subsequent times.

Reply:
my concern was more for if we were changing it into a VbeDllAccessor or whatever, then it would then make more sense to have a standard entry point for the DLL detection rather than having lots of similar DetermineVersionAndExecuteDoEvents, DetermineVersionAndExecuteXYZ... methods with repeated code.   But as you're not doing that... ignore me.